libelement/CMakeLists.txt

```diff
@@ -228,8 +229,13 @@ if (ELEMENT_FORCE_FALLBACK_FS)
     add_definitions(-DELEMENT_FORCE_FALLBACK_FS)
 endif()
 
-add_library(element STATIC ${element_sources})
+add_library(element ${element_sources})
 set_property(TARGET element PROPERTY POSITION_INDEPENDENT_CODE ON)
+target_compile_definitions(element PRIVATE "ELEMENT_EXPORT")
+# make sure we don't add dllimport if building statically
+if (NOT BUILD_SHARED_LIBS)
+    target_compile_definitions(element INTERFACE "ELEMENT_STATIC")
+endif ()
```

Comment:
I might have missed it but I think this might be missing hidden symbol visibility? you can change it with `set_target_properties` and `CXX_VISIBILITY_PRESET` and `C_VISIBILITY_PRESET` to hidden, so that linux/mac don't always export the symbols regardless?

Reply:
I think at the moment we set that via the `#define`s in `common.h`, but it's true that at the moment for things set to `ELEMENT_STATIC` we'd be setting them to default visibility regardless, which isn't right - the static library might then get built into a DLL where you don't want the symbols being exported. I'll change the #defines.

Reply:
I think it would also be an issue with a shared library containing symbols you didn't mark for export? not sure :b but yeah the static library thing is a good point

Reply:
(I'm not sure if this was covered by the slack message?) if you build it as shared library on linux, and it contains symbols that are not marked for export (for whatever reason, they don't have the API macro) then it will still export them, as the default visibility on linux is public

Reply:
Ah I see... I think that's fine? Ultimately what I think we should do is make sure the user _can_ control what gets exported from their code, which this would allow them to do (by ensuring that if libelement was built statically, we aren't applying any visibility attributes to our API functions when they're included by the dylib code).

Essentially - if the dylib author is bothering to care about visibility then we shouldn't get in their way by having opinions about our functions, but if they aren't bothering to care about visibility then... well, they aren't bothering to care about visibility 😛

Reply:
mm maybe 🤔 it feels a bit odd to me that we'd be explicit in marking symbols for export, but leave un-marked symbols as up-in-the-air depending on the consumer and how they configure visibility per-OS

` if they aren't bothering to care about visibility`

the user has to make sure they apply a default hidden visibility globally though, instead of relying on per-target visibility to get it right (i.e `CMAKE_CXX_VISIBILITY_PRESET ` vs `CXX_VISIBILITY_PRESET `. maybe that's normal though? otherwise a static libelement will export it all by default on linux, even if they've set their own targets symbols to be hidden (I don't think it propagates?)

not sure :b it's probably fine, but then libsensation will need updating to apply a global hidden visibility as it only applies to its own target (but then a consumer of libsensation can change that global and unexpectantly change the visibility of libelement, a dependency of a dependency 🤷) so libsensation should really set a per-target visibility setting for libelement regardless in that case (and despite linking to libelement statically, on linux all of its symbols would be visibile in the libsensation DLL that gets built, I think?)

I don't think we want libelement symbols in the libsensation DLL? although it's unlikely you'd want to use a different version of libelement alongside libsensation. I guess it means any breaking changes to libelement's API would then cause a breaking change to the libsensation DLL because a user could be relying on those symbols, and that could be surprising.

Reply:
this would have been an issue before these changes tho, so I'm not sure if I'm wrong in how symbol exporting of a static library works when used in a shared library with a default visibility, or if it's just always been broken?

Reply:
I'm happy to resolve this for the MR but I'm curious what the answer is :b

Reply:
Isn't this unchanged from what we've been doing anyway up until now when building it as static, though? I guess we've just been leaking symbols in static-built libelement forever, which is unfortunate. :smile:

We need the export symbol marking at least for Windows, and it'd seem odd to have it do nothing on ELF platforms. I guess we should just set `CXX_VISIBILITY_PRESET=hidden` on libelement targets (static or shared) and keep the rest as-is?

Reply:
yeah that's how I set it up for libsensation, but I'm not sure if that's better than what you're doing? 🤷 and trying to find any good articles on this stuff is a pain 😢

Reply:
I'll change the target(s) to use hidden symbol visibility as it _should_ be fine with these changes.